vncviewer/touch.cxx

```diff
@@ -0,0 +1,324 @@
+    break;
+  }
+
+  // Run the normal handlers
+  return DefSubclassProc(hWnd, uMsg, wParam, lParam);
+}
+#elif !defined(WIN32) && !defined(__APPLE__)
+static void x11_change_touch_ownership(bool enable)
+{
+  HandlerMap::const_iterator iter;
+
+  XIEventMask *curmasks;
+  int num_masks;
+
+  XIEventMask newmask;
+  unsigned char mask[XIMaskLen(XI_LASTEVENT)] = { 0 };
+
+  newmask.mask = mask;
+  newmask.mask_len = sizeof(mask);
+
+  for (iter = handlers.begin(); iter != handlers.end(); ++iter) {
+    curmasks = XIGetSelectedEvents(fl_display, iter->first, &num_masks);
+    if (curmasks == NULL) {
+      if (num_masks == -1)
+        vlog.error(_("Unable to get X Input 2 event mask for window 0x%08lu"), iter->first);
+      continue;
+    }
+
+    // Our windows should only have a single mask, which allows us to
+    // simplify all the code handling the masks
+    if (num_masks > 1) {
+      vlog.error(_("Window 0x%08lu has more than one X Input 2 event mask"), iter->first);
+      continue;
+    }
+
+    newmask.deviceid = curmasks[0].deviceid;
+
+    assert(newmask.mask_len >= curmasks[0].mask_len);
+    memcpy(newmask.mask, curmasks[0].mask, curmasks[0].mask_len);
+    if (enable)
+      XISetMask(newmask.mask, XI_TouchOwnership);
+    else
+      XIClearMask(newmask.mask, XI_TouchOwnership);
+
+    XISelectEvents(fl_display, iter->first, &newmask, 1);
+
+    XFree(curmasks);
+  }
+}
+
+int x11_grab_pointer(Window window)
+{
+  int status;
+
+  if (handlers.count(window) == 0) {
+    vlog.error(_("Invalid window 0x%08lu specified for pointer grab"), window);
+    return BadWindow;
+  }
+
+  // We need to remove XI_TouchOwnership from our event masks while
+  // grabbing as otherwise we will get double events (one for the grab,
+  // and one because of XI_TouchOwnership) with no way of telling them
+  // apart. See XInputTouchHandler constructor for why we use this
+  // event.
+  x11_change_touch_ownership(false);
+
+  status = dynamic_cast<XInputTouchHandler*>(handlers[window])->grabPointer();
+
+  if (status != XIGrabSuccess)
+    x11_change_touch_ownership(true);
+
+  return status;
+}
+
+void x11_ungrab_pointer(Window window)
+{
+  if (handlers.count(window) == 0) {
+    vlog.error(_("Invalid window 0x%08lu specified for pointer grab"), window);
+    return;
+  }
+
+  dynamic_cast<XInputTouchHandler*>(handlers[window])->ungrabPointer();
+
+  // Restore XI_TouchOwnership now that the grab is gone
+  x11_change_touch_ownership(true);
+}
+#endif
+
+#if !defined(WIN32) && !defined(__APPLE__)
+extern void dbgEvent(const XEvent& event);
+#endif
+
+static int handleTouchEvent(void *event, void *data)
+{
+#if defined(WIN32)
+  MSG *msg = (MSG*)event;
+
+  // Don't proceed if this isn't a touch device
+  if (!GetSystemMetrics(SM_DIGITIZER)) {
+    return 0;
+  }
+
+  // If window is registered as touch we can not receive gestures
+  if (IsTouchWindow(msg->hwnd, NULL)) {
+    vlog.error(_("Window is registered for touch instead of gestures"));
```

Comment:
Won't this spam lots of errors?

Reply:
This will only happen if the window was explicitly registered as a touch window. In my mind that can only happen due to a programming error, so with 62546c58622e5706b1a79a30fae6c28e35f381c3 we now exit the viewer when this happens. 

Note that this code was further modified in 7fcb1048d6292357c133b93b93d6c1fa03a97ad6 later.